### PR TITLE
Include soak tests

### PR DESCRIPTION
**Pull request by zpqrtbnk**
This PR includes the soak tests in the repository, as an example. This means that the soak tests are, in fact, examples just like the other examples. Once you clone the repository, you can build it all with:
```
./hz.ps1 build
```
and then you can run the `Hazelcast.Examples.SoakTests.Soak1` soak test with 1 thread and for 1 minute, on two localhost members, with:
```
./hz.ps1 run-example ~Soak1 --%
  --hazelcast:examples:soak1:ThreadCount=1
  --hazelcast:examples:soak1:Duration="00:01:00"
  --hazelcast:networking:addresses:0=127.0.0.1:5701
  --hazelcast:networking:addresses:1=127.0.0.1:5702
```

I am also working on getting GitHub Actions to build the examples for you, so that you would just have to download an artifact from GitHub and run the soak test with:
```
./hx.exe `Soak1
  --hazelcast:examples:soak1:ThreadCount=1
  --hazelcast:examples:soak1:Duration="00:01:00"
  --hazelcast:networking:addresses:0=127.0.0.1:5701
  --hazelcast:networking:addresses:1=127.0.0.1:5702
```
So things will be even easier - but that will be for the next release.

However, the plumbing to do it already exists, and may actually be useful now if you need to package the examples / soak tests in order to copy them to some remote machine where you don't want to *build* them. After you have built the examples, you can do:
```
./hz.ps1 publish-examples
```
And that will create the `temp/publish` directory, containing one sub-directory per target (`net462`, `netcoreapp2.1` and `netcoreapp3.1`) with everything you need to run the examples. So, say you want to run the `netcoreapp3.1` code on a remote machine, you only need to zip and copy over that directory. Simple.

Review: ensure you can clone the repository, build, and run the soak test for some amount of time. Ensure you can package (with `publish-examples`) the tests, copy them to some other place, and run them.

Note: this is the original soak test we had with v3, and it's pretty crude. It just puts and gets values in a map, but does not test what happens when a member dies, etc. Probably we will want to create better, more complete, tests.

### CODE REVIEW

**Comment by zpqrtbnk** [comment]
And, finally, you may want to update the Confluence .NET Release Document with whatever doc you think is important.

File: src/Hazelcast.Net.Examples/SoakTests/Soak1.cs [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+                .Build();
+
+            var logger = options.LoggerFactory.Service.CreateLogger<Soak1>();
+
+            // say hello
+            logger.LogInformation("Hazelcast .NET soak tests");
+
+            // get duration
+            var duration = TimeSpan.FromMinutes(1); // default to 1 minute
+            if (soakOptions.Duration.EndsWith("h"))
```
Comment (alexb5dh) [inline src/Hazelcast.Net.Examples/SoakTests/Soak1.cs:71]: `soakOptions.Duration` is `null` by default so we get `NullReferenceException` if run without arguments.

File: src/Hazelcast.Net.Examples/SoakTests/Soak1.cs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Hazelcast.DistributedObjects;
+using Hazelcast.Examples.Models;
+using Hazelcast.Query;
+using Microsoft.Extensions.Logging;
+
+namespace Hazelcast.Examples.SoakTests
+{
+    // ./hz.ps1 run-example ~Soak1 --%
```
Comment (alexb5dh) [inline src/Hazelcast.Net.Examples/SoakTests/Soak1.cs:27]: Is  `--%` needed if commands ends with it?

**Review by alexb5dh — COMMENTED** [review]
Unfortunately never got
```
./hz.ps1 run-example ~Soak1 --%
  --hazelcast:examples:soak1:ThreadCount=1
  --hazelcast:examples:soak1:Duration="00:01:00"
  --hazelcast:networking:addresses:0=127.0.0.1:5701
  --hazelcast:networking:addresses:1=127.0.0.1:5702
```
to work. Did it pass parameters correctly for you?  

For me this gives 
```
ParserError:
Line |
   2 |    --hazelcast:examples:soak1:ThreadCount=1
     |      ~
     | Missing expression after unary operator '--'.
```

Tried different combinations with backtick and `--%` but PS either complains or forwards parameters to test in the wrong format.

Only launching it manually via hx.exe from command line worked.

File: src/Hazelcast.Net.Examples/SoakTests/Soak1.cs (hunk 1 above)
Comment (zpqrtbnk) [reply src/Hazelcast.Net.Examples/SoakTests/Soak1.cs:95]: fixing that one, `soakOptions.Duration` will be `1m` (ie 1 minute) by default

File: src/Hazelcast.Net.Examples/SoakTests/Soak1.cs (hunk 2 above)
Comment (zpqrtbnk) [reply src/Hazelcast.Net.Examples/SoakTests/Soak1.cs:29]: as in, if we run without arguments at all? no, not needed - that's a PowerShell thing, PS is super-annoying with arguments

**Comment by zpqrtbnk** [comment]
so... 
```
./hz.ps1 run-example ~Soak1 --%
  --hazelcast:examples:soak1:ThreadCount=1
  --hazelcast:examples:soak1:Duration="00:01:00"
  --hazelcast:networking:addresses:0=127.0.0.1:5701
  --hazelcast:networking:addresses:1=127.0.0.1:5702
```
was misleading - this is seen as 5 independent command lines by PowerShell, which therefore tries to execute `--hazelcast:examples:soak1:ThreadCount=1` as if it were a command. if you start this from the PowerShell prompt, you need to end each line with the PowerShell continuation character ` (backtick):
```
./hz.ps1 run-example ~Soak1 --% `
  --hazelcast:examples:soak1:ThreadCount=1 `
  --hazelcast:examples:soak1:Duration="00:01:00" `
  --hazelcast:networking:addresses:0=127.0.0.1:5701 `
  --hazelcast:networking:addresses:1=127.0.0.1:5702
```

because I run from bash, I use bash continuation character \ (backslash):

```
./hz.sh run-example ~Soak1 --% \
  --hazelcast:examples:soak1:ThreadCount=1 \
  --hazelcast:examples:soak1:Duration="00:01:00" \
  --hazelcast:networking:addresses:0=127.0.0.1:5701 \
  --hazelcast:networking:addresses:1=127.0.0.1:5702
```

File: src/Hazelcast.Net.Examples/SoakTests/Soak1.cs [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    //
+    // ./hx.exe ~Soak1
+    //   --hazelcast:examples:soak1:ThreadCount=1             (valid integer)
+    //   --hazelcast:examples:soak1:Duration="HH:MM:SS"       (valid timespan)
+    //   --hazelcast:networking:addresses:0=127.0.0.1:5701
+    //   --hazelcast:networking:addresses:1=127.0.0.1:5702
+    //
+    // each server must run a 4.2+ server; start a server once by running:
+    //
+    // ./hz.ps1 server -server 4.2
```
Comment (alisengul53) [inline src/Hazelcast.Net.Examples/SoakTests/Soak1.cs:54]: Should be "./hz.ps1 run-server -server 4.2"?

Comment (zpqrtbnk) [reply src/Hazelcast.Net.Examples/SoakTests/Soak1.cs:54]: totally, fixing

**Comment by alexb5dh** [comment]
> so...
> 
> ```
> ./hz.ps1 run-example ~Soak1 --%
>   --hazelcast:examples:soak1:ThreadCount=1
>   --hazelcast:examples:soak1:Duration="00:01:00"
>   --hazelcast:networking:addresses:0=127.0.0.1:5701
>   --hazelcast:networking:addresses:1=127.0.0.1:5702
> ```
> 
> was misleading - this is seen as 5 independent command lines by PowerShell, which therefore tries to execute `--hazelcast:examples:soak1:ThreadCount=1` as if it were a command. if you start this from the PowerShell prompt, you need to end each line with the PowerShell continuation character ` (backtick):
> 
> ```
> ./hz.ps1 run-example ~Soak1 --% `
>   --hazelcast:examples:soak1:ThreadCount=1 `
>   --hazelcast:examples:soak1:Duration="00:01:00" `
>   --hazelcast:networking:addresses:0=127.0.0.1:5701 `
>   --hazelcast:networking:addresses:1=127.0.0.1:5702
> ```
> 
> because I run from bash, I use bash continuation character \ (backslash):
> 
> ```
> ./hz.sh run-example ~Soak1 --% \
>   --hazelcast:examples:soak1:ThreadCount=1 \
>   --hazelcast:examples:soak1:Duration="00:01:00" \
>   --hazelcast:networking:addresses:0=127.0.0.1:5701 \
>   --hazelcast:networking:addresses:1=127.0.0.1:5702
> ```

Unfortunately backtick seems to be ignored after `

**Comment by zpqrtbnk** [comment]
oh my - PowerShell is powerfully annoying with args, and I *really* though I had fixed it - can you list the various things you've tried and the corresponding error? that'd help me a lot

File: src/Hazelcast.Net.Examples/SoakTests/Soak1.cs [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    //   --hazelcast:examples:soak1:Duration="HH:MM:SS"       (valid timespan)
+    //   --hazelcast:networking:addresses:0=127.0.0.1:5701
+    //   --hazelcast:networking:addresses:1=127.0.0.1:5702
+    //
+    // each server must run a 4.2+ server; start a server once by running:
+    //
+    // ./hz.ps1 run-server -server 4.2
+    //
+    // this will download all the required JARs from the proper repositories, *and*
+    // a default hazelcast-<version>.xml configuration file. stop the server, and
```
Comment (alisengul53) [inline src/Hazelcast.Net.Examples/SoakTests/Soak1.cs:57]: Would be nice to remark the path. /build/hazelcast-<version>.xml

File: .github/workflows/build-release.yml [hunk 5]
```diff
@@ -263,6 +263,20 @@ jobs:
           name: NuGet-Packages (${{ matrix.os }})
           path: './temp/output/Hazelcast.Net.*pkg'
 
+      # publish examples (inc. soak tests) (windows only so we have framework too)
+      - name: Publish examples
+        if: matrix.os == 'windows-latest'
+        shell: pwsh
+        run: ./hz.ps1 -noRestore -localRestore publish-examples
```
Comment (alexb5dh) [inline .github/workflows/build-release.yml:270]: As I understand for now this command is not yet present in the PR? 
`./hz.ps1 publish-examples` fails with `unknown command 'publish-examples' - use 'help' to list valid commands`

File: src/Hazelcast.Net.Examples/SoakTests/Soak1.cs (hunk 4 above)
Comment (zpqrtbnk) [reply src/Hazelcast.Net.Examples/SoakTests/Soak1.cs:57]: yes, adding

File: src/Hazelcast.Net.Examples/SoakTests/Soak1.cs [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+            while (!token.IsCancellationRequested)
+            {
+                try
+                {
+                    map = await client.GetMapAsync<string, string>("soak1").ConfigureAwait(false);
+                    break;
+                }
+                catch (Exception e)
+                {
+                    logger.LogError(e.StackTrace);
```
[the remark this answers is not in the document]
Comment (zpqrtbnk) [reply src/Hazelcast.Net.Examples/SoakTests/Soak1.cs:186]: no, fixing

File: .github/workflows/build-release.yml (hunk 5 above)
Comment (zpqrtbnk) [reply .github/workflows/build-release.yml:270]: but? should be part of the PR, fixing

Comment (zpqrtbnk) [reply .github/workflows/build-release.yml:270]: so, it disapeared in a poorly-controled git-rebase, now it's back - publishes to `temp/examples` - in theory you should be able to eg `cd temp/examples/netcoreapp3.1` and there, `hx.exe ...` should run the examples. also producing a zipped version so it's easy to deploy to, say, soak test machine.

**Review by alisengul53 — APPROVED** [review]

**Review by alexb5dh — APPROVED** [review]

**Pull request merged**